### ISSUE TITLE
Support IdelingResource (Espresso)

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-java#4190]
When UI testing using Espresso, it can automatically detect when all async work has completed. This way it knows when the system is "stable" so it can query the UI state. This is done using a concept known as `IdelingResource`: https://developer.android.com/reference/android/support/test/espresso/IdlingResource.html

To ease Espresso UI testing with Realm we should expose our internal thread pools in such a way that they can be made compatible with this interface. 

I can see us doing it in 2 steps:

1) Make our executors public somehow, either through public fields or getters. Not sure how  #3834 will effect this. @beeender ?

2) Provide some copy-paste snippet code on the website showing how to use it. 

Implementing `IdelingResource` directly ourselves is probably not possible as we do not want to depend on the Espresso dependency outside testing.

Example of this being used here:
http://blog.sqisland.com/2015/04/espresso-custom-idling-resource.html
https://github.com/chiuki/espresso-samples/blob/master/idling-resource-intent-service/app/src/androidTest/java/com/sqisland/espresso/idling_resource/intent_service/IntentServiceIdlingResource.java
https://github.com/chiuki/espresso-samples/blob/master/idling-resource-intent-service/app/src/androidTest/java/com/sqisland/espresso/idling_resource/intent_service/MainActivityTest.java

**Comment by Zhuinden** [comment]
https://github.com/AzimoLabs/ConditionWatcher this is much more reliable than Espresso imo

**Comment by beeender** [comment]
After #3834, only the async transaction will be using executors, and from the executor perspective, it will have no difference compared with before.

**Comment by cmelchior** [comment]
So for us to support this I would assume we need to expose information from the ObjectStore worker thread somehow?

**Comment by beeender** [comment]
yeah, but i am not quite sure if it is doable. The "worker thread" in the Object Store is actually a daemon thread which lives always. And it is waiting on `epoll` events when it is in idle.
